File: src/main/java/com/senderman/stickersorterbot/web/controller/MainPageController.java
package com.senderman.stickersorterbot.web.controller;

import com.annimon.tgbotsmodule.services.CommonAbsSender;
import com.senderman.stickersorterbot.model.Sticker;
import com.senderman.stickersorterbot.repository.StickerRepository;
import com.senderman.stickersorterbot.util.BotExtensions;
import com.senderman.stickersorterbot.web.CachingStickerFileProvider;
import com.senderman.stickersorterbot.web.entity.WebSticker;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;

import java.security.Principal;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

@Controller
public class MainPageController {

    private final String cacheDir;
    private final CommonAbsSender telegram;
    private final StickerRepository stickerRepository;
    private final CachingStickerFileProvider stickerCache;

    public MainPageController(
            @Value("${website.cacheDirectory}") String cacheDir,
            CommonAbsSender telegram,
            StickerRepository stickerRepository,
            CachingStickerFileProvider stickerCache
    ) {
        this.cacheDir = cacheDir;
        this.telegram = telegram;
        this.stickerRepository = stickerRepository;
        this.stickerCache = stickerCache;
    }

    @GetMapping("/")
    public String showMainPage(
            @RequestParam(value = "search", required = false, defaultValue = "") String search,
            Principal principal,
            Model model
    ) {

        int userId = Integer.parseInt(principal.getName());
        String[] tagsToFind = search.split("\\s+");

        var stickers = search.isBlank() ?
                stickerRepository.findAllByUserId(userId) :
                stickerRepository.findAllByUserIdAndTagsContaining(userId, List.of(tagsToFind));

        generateWebContent(userId, model, stickers);
        return "main";
    }

    private void generateWebContent(int userId, Model model, Collection<Sticker> source) {

        String userName = BotExtensions.getFirstNameById(userId, telegram);
        model.addAttribute("username", userName);

        var webStickers = mapToWebStickers(source);

        model.addAttribute("stickers", webStickers);
    }

    private Collection<WebSticker> mapToWebStickers(Collection<Sticker> stickers) {
        return stickers.stream().map(it -> {
            var stickerFile = stickerCache.retrieveSticker(it);
            return new WebSticker(it, "/" + cacheDir + "/" + stickerFile.getName());
        }).collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/senderman/stickersorterbot/web/controller/MainPageController.kt
package com.senderman.stickersorterbot.web.controller

import com.annimon.tgbotsmodule.services.CommonAbsSender
import com.senderman.stickersorterbot.bot.getFirstNameById
import com.senderman.stickersorterbot.model.Sticker
import com.senderman.stickersorterbot.model.StickerRepository
import com.senderman.stickersorterbot.web.CachingStickerFileProvider
import com.senderman.stickersorterbot.web.entities.WebSticker
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestParam
import java.security.Principal

@Controller
class MainPageController(
        @Value("\${website.cacheDirectory}")
        private val cacheDir: String,

        private val telegram: CommonAbsSender,
        private val stickerRepo: StickerRepository,
        private val stickerCache: CachingStickerFileProvider
) {

    @GetMapping("/")
    fun showMainPage(
            @RequestParam("search", required = false, defaultValue = "") search: String,
            principal: Principal,
            model: Model
    ): String {
        val userId = principal.name.toInt()
        val stickers = stickerRepo.findAllByUserId(userId)

        // filter by search query
        /* if (!search.isBlank()) {
             val tagsToFind = search.split(Regex("\\s+"))
             stickers.removeAll { it.name !in tagsToFind }
         }*/

        return generateWebContent(userId, model, stickers)
    }

    @PostMapping("/deleteSticker")
    fun deleteSticker(
            @RequestParam("search", required = false, defaultValue = "") search: String,
            fileUniqueId: String,
            principal: Principal,
            model: Model
    ): String {
        val userId = principal.name.toInt()
        stickerRepo.deleteById(Sticker.generateId(userId, fileUniqueId))
        return showMainPage(search, principal, model)
    }

    // generate web page's content from user's stickers or from given source, if present
    private fun generateWebContent(userId: Int, model: Model, source: MutableSet<Sticker>): String {

        val username = telegram.getFirstNameById(userId)
        model.addAttribute("username", username)

        val webStickers: Iterable<WebSticker> = mapToWebStickers(source)

        model.addAttribute("stickers", webStickers)
        return "main"
    }

    private fun mapToWebStickers(stickers: Iterable<Sticker>): Iterable<WebSticker> =
            stickers.map {
                val stickerFile = stickerCache.retrieveSticker(it)
                WebSticker(it, "/$cacheDir/${stickerFile.name}")
            }
}